### DISCUSSION TITLE
Nuxt3 node .output/server/index.mjs  不能公网IP访问

### DISCUSSION THREAD

**Discussion by shenghy** [nuxt/nuxt#17721] [Questions]
## Nuxt3 node .output/server/index.mjs  不能公网IP访问
> I'm very sorry, my English is not good, so I use Chinese to describe it. I hope what I can understand can help me.
非常抱歉，我英文不好，所以用中文描述，非常抱歉。希望能看懂的可以帮助我一下。

我的环境和版本如下：

`package.json文件`

```PowerShell
{
  "private": true,
  "scripts": {
    "dev": "nuxi dev --port 8080",
    "build": "nuxi build",
    "start": "node .output/server/index.mjs "
  },
  "devDependencies": {
    "babel-plugin-transform-remove-console": "^6.9.4",
    "nuxt3": "^3.0.0-27424151.dbab979"
  },
  "dependencies": {
    "default-passive-events": "^2.0.0",
    "highlight.js": "^11.4.0"
  }
}
```

服务器环境：Linux CentOS 8.x

我项目开发完成后，想部署的时遇到了问题。问题如下：

使用命令打包后，把`.output`文件夹上传到服务器上，然后用用` node `  开启服务时，显示了服务成功，也可用  `http://127.0.0.1:3000` 和 `http://localhost:3000`访问到，但是当用 公网ip访问时就访问不到。比如用 http://191.36.81.61:3000/ 就访问不到。

```PowerShell
$ yarn build
$ yarn start
yarn run v1.22.15
$ node .output/server/index.mjs 
Listening on http://localhost:3000/
curl http://localhost:3000/    // 可以访问到
curl http://191.36.61.61:3000/  //访问不到

```

我可以保证对外的端口和安全设置是没有问题的，因为我同样在3000端口开启其它服务，是可以通过外网访问的。

我在windows上也可以复现这个问题，就是用 局域网IP无法访问，例如：http://192.168.0.99:3000/(不可以访问)

我觉的应该是我哪里配置不正确，或者部署的时的操作有问题。在网上找了一天也不知道如何解决。正因为我的英文不好，所以我在官网上的文档中也没有得到很好的解决。希望有知道如何解决的大神可以帮忙告知一下。

**Comment by ghostchan** [comment]
胖哥，试过这种没有，https://segmentfault.com/q/1010000009999934，代码仓库也可以贴出来，大家试试

**Comment by shenghy** [comment]
@ghostchan 你给的这个答案是Nuxt2版本的，在Nuxt3里，我这样配置不起作用。

**Reply by ghostchan** [reply]
好的，后边你要解决了，到时候分享一下，在这顺便吐个槽哈，你的新博客，感觉没有旧的好看，我用nuxt仿照你之前的博客搞了一个，哈哈    https://www.happyfe.com

**Comment by zhixXiang** [comment]
我也有遇到這個問題，目前已經順利解決
目前是使用pm2: https://v3.nuxtjs.org/docs/deployment/pm2

而使用pm2的話可以設定environment variables
https://pm2.keymetrics.io/docs/usage/environment/

env底下設定PORT: 80 和 HOST: "0.0.0.0" 即可順利訪問!

```
module.exports = {
  apps: [
    {
      name: 'NuxtApp',
      exec_mode: 'cluster',
      instances: 'max',
      script: '.output/server/index.mjs',
      env: {
            "PORT": 80,
            "HOST": "0.0.0.0"
      }
    }
  ]
}
```

**Reply by valetzx** [reply]
你好，这个env要写在哪个文件里。没接触过 node vue nuxt ）根别人教程部署遇到类似问题

**Reply by zhixXiang** [reply]
自己建立一個ecosystem.config.js(建在和.output同一層目錄)，然後寫在裡面
再用pm2跑起來就可以了
```
pm2 start ecosystem.config.js
```

**Reply by valetzx** [reply]
好的感谢

**Comment by valetzx** [comment]
请问解决了吗？有教程吗感谢！
